src/Security/MemberAuthenticator/CookieAuthenticationHandler.php

```diff
@@ -0,0 +1,248 @@
+                $now = DBDatetime::now();
+                $now = new \DateTime($now->Rfc2822());
+                if ($now > $expiryDate) {
+                    $member = null;
+                }
+            }
+        }
+
+        if ($member) {
+            if ($this->cascadeLogInTo) {
+                // @todo look at how to block "regular login" triggers from happening here
+                // @todo deal with the fact that the Session::current_session() isn't correct here :-/
+                $this->cascadeLogInTo->logIn($member, false, $request);
+            }
+
+            // @todo Consider whether response should be part of logIn() as well
+
+            // Renew the token
+            if ($rememberLoginHash) {
+                $rememberLoginHash->renew();
+                $tokenExpiryDays = RememberLoginHash::config()->uninherited('token_expiry_days');
+                Cookie::set(
+                    $this->getTokenCookieName(),
+                    $member->ID . ':' . $rememberLoginHash->getToken(),
+                    $tokenExpiryDays,
+                    null,
+                    null,
+                    false,
+                    true
+                );
+            }
+
+            // Audit logging hook
+            $member->extend('memberAutoLoggedIn');
+
+            return $member;
+        }
+    }
+
+    /**
+     * @param Member $member
+     * @param bool $persistent
+     * @param HTTPRequest $request
+     * @return \SilverStripe\Control\HTTPResponse|void
+     */
+    public function logIn(Member $member, $persistent = false, HTTPRequest $request = null)
+    {
+        // @todo couple the cookies to the response object
+
+        // Cleans up any potential previous hash for this member on this device
+        if ($alcDevice = Cookie::get($this->getDeviceCookieName())) {
+            RememberLoginHash::get()->filter('DeviceID', $alcDevice)->removeAll();
+        }
+
+        // Set a cookie for persistent log-ins
+        if ($persistent) {
+            $rememberLoginHash = RememberLoginHash::generate($member);
+            $tokenExpiryDays = RememberLoginHash::config()->uninherited('token_expiry_days');
+            $deviceExpiryDays = RememberLoginHash::config()->uninherited('device_expiry_days');
+            Cookie::set(
+                $this->getTokenCookieName(),
+                $member->ID . ':' . $rememberLoginHash->getToken(),
+                $tokenExpiryDays,
+                null,
+                null,
+                null,
+                true
+            );
+            Cookie::set(
+                $this->getDeviceCookieName(),
+                $rememberLoginHash->DeviceID,
+                $deviceExpiryDays,
+                null,
+                null,
+                null,
+                true
+            );
+
+        } else {
+            // Clear a cookie for non-persistent log-ins
+            $this->clearCookies();
```

Comment:
Why does logIn() call logOut()?
tractorcow 2 days ago Contributor

If the intent is to clear cookies, take the logic out of logOut() into a method that does just that, and invoke it from here. Code intent smells bad as it stands.